Polish schema errors

DennisDenuto opened the pull request:
fixes #396 

Introduces a `SchemaError` that accepts errors with the goal of producing a consistent error message. Kinda like a report of all the errors it knows about.

`SchemaError` will always print a 'summary' when `Error` is called.

If `SchemaError` is given an error of type `schemaAssertionError` it will report those errors in a 'diff' format. with the line, file name, and expected / actual value formatted in a nice readable way.

For all other type of errors given to `SchemaError`, it will delegate to their `Error()`  and append those messages to the end of the `SchemaError` 'report'

DennisDenuto committed:
Refactor: unexport fields on schema error

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
UX: Print out multiple hints during invalid schema

- invalidSchemaError now can be provided multiple hints

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
UX: Improve schema error messages

- Change the output of the error message to hopefully make it clearer
what the problem is, and where in the schema file the problem is too.
- Use text/template package to generate error message. Hope it is
simpler to grok what the error message will render as, and easier to
make changes vs relying on multiple Printf commands

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
UX: Remove redundant info during schema error

- When printing the filename, stop printing the line number, since the
line number is printed on the next section (in the 'diff' section)

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto commented:
As per the instructions: Closing to re-sign the CLA and will re-open this PR

DennisDenuto committed:
UX: Improve schema error array definition message

- Use the invalid schema template to highlight the line numbers and how
to fix the error. This makes it consistent with how other schema errors
will look.

Authored-by: Dennis Leon <leonde@vmware.com>

pivotaljohn commented on pkg/schema/error.go:
```diff
@@ -4,25 +4,47 @@
-		Message: message,
-		Found:   found,
-		Hint:    hint,
+const invalidSchemaTemplate = `
+{{.Title}}
+
+{{.FileName}}:
+{{pad "|" false}}
+{{pad "|" true}} {{.Diff}}
+{{pad "|" false}}
+
+{{with .Found}}{{pad "=" false}} found: {{.}}{{end}}
+{{with .Expected}}{{pad "=" false}} expected: {{.}}{{end}}
+{{range .Hints}}{{pad "=" false}} hint: {{.}}
+{{end}}`
```
I dig it:
- 👍🏻  delegate formatting to stdlib
- 👍🏻  declarative, concise, and easy to edit.

pivotaljohn commented on pkg/schema/error.go:
```diff
@@ -40,47 +62,49 @@ func NewUnexpectedKeyError(found *yamlmeta.MapItem, definition *filepos.Position
-	Hint    string
+	Title    string
+	FileName string
+	Diff     string
```
If I'm reading the usage right, I think `Diff` is the original source.... 

```suggestion
	Source   string
```

?

DennisDenuto replied:
Good callout. done!

DennisDenuto committed:
Refactor:

- Generalize invalidSchemaError to schemaError.
-- schemaError will be used to produce a consistent error across the
different error scenarios (not limited to an invalid schema file)

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
UX: Improve schema error when type mismatch occurs

- SchemaError now has an array of schemaAssertionErrors since a type
mismatch can span multiple key/values. Hence when building the schema
error report, allows the ability to display the various 'errors'
associated.

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
UX: Improve schema error when unexpected key is provided

 - This case is currently not yet possible to run into, until the
 following issue is resolved: https://github.com/vmware-tanzu/carvel-ytt/issues/344

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Refactor

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
UX: Add a summary section when reporting schema errors

- The first line in the schema error will be a summary of the error/s
- Also allow for descriptions (optional) to also be part of each assertion
failure

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Stop skipping tests in schema_test.go

- These tests do *not* have a data value file, and hence are missing
line numbers and file names. This results in a less-than-ideal output in
the error message. Hence why they were skipped until a *better* output
message could be figured out.
-- However, now the plan is to split off that 'improving' error messages for
such above cases into a separate GH issue. And is out of scope for:
https://github.com/vmware-tanzu/carvel-ytt/issues/396

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Refactor

- minor re-wording of error messages

Authored-by: Dennis Leon <leonde@vmware.com>

pivotaljohn commented:
@DennisDenuto

Regarding the markdown style header line...
... it didn't land for me at first.
... but then you shared with me the intent of looking at this output as a kind of report. That really made a shift in my head.

I want to see this experiment get some air.

Thanks for your thoughtful approach to this story.
